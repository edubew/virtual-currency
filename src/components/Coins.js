// import React, { useEffect, useState } from 'react';
// import { useDispatch, useSelector } from 'react-redux';
// import { BiSearch } from 'react-icons/bi';
// import { coinsEffect } from '../redux/reducers/coinsReducers';

// const Coins = () => {
//   const dispatch = useDispatch();
//   const coins = useSelector((state) => state.coins.coinsState);
//   const [search, setSearch] = useState('');

//   useEffect(() => {
//     dispatch(coinsEffect);
//   }, [dispatch]);

//   const filteredCoins = coins
//     ? coins.filter((coin) => coin.name.toLowerCase().includes(search.toLowerCase()))
//     : [];

//   return (
//     <div>
//       <div className="search__container">
//         <span><BiSearch /></span>
//         <input
// className="search__input"
// type="text"
// placeholder="search..."
// onChange={(e) => setSearch(e.target.value)}
// autoComplete="off" />
//       </div>
//       <div className="cards__container">
//         {filteredCoins.map((coin) => (
//           <article key={coin.id}>
//             <div className="img__symbol">
//               <img src={coin.image} alt="avator" />
//               <p>{coin.symbol}</p>
//             </div>
//             <h3>{coin.name}</h3>
//             <p>{coin.market_cap}</p>
//           </article>
//         ))}
//       </div>
//     </div>
//   );
// };

// export default Coins;

import React, { useEffect, useState } from 'react';
import axios from 'axios';
// import { useDispatch } from 'react-redux';
import { BiSearch } from 'react-icons/bi';
// import { GET_COINS, GET_COIN_DETAILS } from '../redux/actions/coinActions';

const Coins = () => {
  const url = 'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=20&page=1&sparkline=false';
  // const dispatch = useDispatch();
  const [coins, setCoins] = useState([]);
  // const [data, setData] = useState([]);
  const [search, setSearch] = useState('');
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  useEffect(() => {
    const fetchCoins = async () => {
      try {
        const response = await axios.get(url);
        setCoins(response.data);
        setLoading(false);
      } catch (error) {
        setError(error);
        setLoading(false);
      }
    };
    fetchCoins();
  }, [url]);
  // const handleCoinClick = async (id) => {
  //   try {
  //     const response = await axios.get(`${url}/${id}/market_charts&limit=20`);
  //     setData(response.data);
  //     dispatch({ type: GET_COIN_DETAILS, payload: response.markets_charts });
  //   } catch (error) {
  //     console.error(`Error fetching coin details for id ${id}:`, error);
  //   }
  // };
  // onClick={() => handleCoinClick(coin.id)}
  const filteredCoins = coins.filter((coin) => coin.name.toLowerCase().includes(search.toLowerCase()));
  const coinList = filteredCoins.map((coin) => (
    <div key={coin.id}>
      <article>
        <img src={coin.image} alt="coin" />
        <p>{coin.market_cap_rank}</p>
        <p>{coin.name}</p>
        <p>{coin.market_cap}</p>
      </article>
    </div>
  ));
  return (
    <div>
      <div className="search">
        <BiSearch className="search-icon" />
        <input
          type="text"
          placeholder="Search"
          className="search-input"
          onChange={(e) => setSearch(e.target.value)}
        />
      </div>
      {error ? (
        <p>{error.message}</p>
      ) : loading ? (
        <p>Loading...</p>
      ) : (
        <div className="coin-list">{coinList}</div>
      )}
    </div>
  );
};
export default Coins;
