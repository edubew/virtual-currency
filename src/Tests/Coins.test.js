import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios';
import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import store from '../redux/configureStore';
import Coins from '../components/Coins';

jest.mock('axios');

const responseData = [
  {
    id: 'bitcoin',
    symbol: 'btc',
    name: 'Bitcoin',
    image: 'https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579',
    current_price: 27363,
    market_cap: 529135059039,
    market_cap_rank: 1,
    fully_diluted_valuation: 574162376351,
    total_volume: 18472693106,
    high_24h: 28246,
    low_24h: 27201,
    price_change_24h: -580.4379283007438,
    price_change_percentage_24h: -2.07716,
    market_cap_change_24h: -11095646459.78186,
    market_cap_change_percentage_24h: -2.05387,
    circulating_supply: 19353125,
    total_supply: 21000000,
    max_supply: 21000000,
    ath: 69045,
    ath_change_percentage: -60.38593,
    ath_date: '2021-11-10T14:24:11.849Z',
    atl: 67.81,
    atl_change_percentage: 40236.01248,
    atl_date: '2013-07-06T00:00:00.000Z',
    roi: null,
    last_updated: '2023-04-22T07:37:08.824Z',
  },
];

describe('Test Coins Component', () => {
  it('should fetch data from API when page loads', async () => {
    axios.get.mockResolvedValue({ data: responseData });
    let asFragment;
    await act(() => {
      const component = render(
        <BrowserRouter>
          <Provider store={store}>
            <Coins />
          </Provider>
        </BrowserRouter>,
      );
      asFragment = component.asFragment;
    });
    await screen.findAllByText(responseData[0].name);
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getAllByText(responseData[0].name)).toBeTruthy();
  });
});
