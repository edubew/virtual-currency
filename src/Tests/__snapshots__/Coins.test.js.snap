// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Coins Component should fetch data from API when page loads 1`] = `
<DocumentFragment>
  <section
    class="sc-gswNZR bbHXwx"
  >
    <div
      class="navbar"
    >
      <h1
        class="brand__name"
      >
        <a
          href="/"
        >
          Crypto Space
        </a>
      </h1>
      <form
        class="sc-bcXHqe jEOLbp"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          class="search__input"
          maxlength="50"
          placeholder="Search by name..."
          type="text"
          value=""
        />
      </form>
      <div
        class="nav__icons"
      >
        <svg
          baseProfile="tiny"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          version="1.2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 16c2.206 0 4-1.795 4-4v-6c0-2.206-1.794-4-4-4s-4 1.794-4 4v6c0 2.205 1.794 4 4 4zM19 12v-2c0-.552-.447-1-1-1s-1 .448-1 1v2c0 2.757-2.243 5-5 5s-5-2.243-5-5v-2c0-.552-.447-1-1-1s-1 .448-1 1v2c0 3.52 2.613 6.432 6 6.92v1.08h-3c-.553 0-1 .447-1 1s.447 1 1 1h8c.553 0 1-.447 1-1s-.447-1-1-1h-3v-1.08c3.387-.488 6-3.4 6-6.92z"
          />
        </svg>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416.3 256c0-21 13.1-38.9 31.7-46.1-4.9-20.5-13-39.7-23.7-57.1-6.4 2.8-13.2 4.3-20.1 4.3-12.6 0-25.2-4.8-34.9-14.4-14.9-14.9-18.2-36.8-10.2-55-17.3-10.7-36.6-18.8-57-23.7C295 82.5 277 95.7 256 95.7S217 82.5 209.9 64c-20.5 4.9-39.7 13-57.1 23.7 8.1 18.1 4.7 40.1-10.2 55-9.6 9.6-22.3 14.4-34.9 14.4-6.9 0-13.7-1.4-20.1-4.3C77 170.3 68.9 189.5 64 210c18.5 7.1 31.7 25 31.7 46.1 0 21-13.1 38.9-31.6 46.1 4.9 20.5 13 39.7 23.7 57.1 6.4-2.8 13.2-4.2 20-4.2 12.6 0 25.2 4.8 34.9 14.4 14.8 14.8 18.2 36.8 10.2 54.9 17.4 10.7 36.7 18.8 57.1 23.7 7.1-18.5 25-31.6 46-31.6s38.9 13.1 46 31.6c20.5-4.9 39.7-13 57.1-23.7-8-18.1-4.6-40 10.2-54.9 9.6-9.6 22.2-14.4 34.9-14.4 6.8 0 13.7 1.4 20 4.2 10.7-17.4 18.8-36.7 23.7-57.1-18.4-7.2-31.6-25.1-31.6-46.2zm-159.4 79.9c-44.3 0-80-35.9-80-80s35.7-80 80-80 80 35.9 80 80-35.7 80-80 80z"
          />
        </svg>
      </div>
    </div>
    <a
      href="/coins/bitcoin"
    >
      <article
        class="filter__item"
      >
        <img
          alt="coin"
          src="https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579"
        />
        <div
          class="cotents"
        >
          <p>
            Bitcoin
          </p>
          <p>
            $27363
          </p>
        </div>
      </article>
    </a>
    <h1
      class="title"
    >
      Crypto Data
    </h1>
    <div
      class="coin__grid"
    >
      <div
        class="grid__layout"
      >
        <a
          href="/coins/bitcoin"
        >
          <article>
            <img
              alt="coin"
              src="https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579"
            />
            <p>
              Bitcoin
            </p>
            <p>
              $27363
            </p>
            <p
              class="red"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="17"
                  x2="7"
                  y1="7"
                  y2="17"
                />
                <polyline
                  points="17 17 7 17 7 7"
                />
              </svg>
              -2.07716%
            </p>
          </article>
        </a>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
